### PR TITLE
ENH pip check migration

### LINKED ISSUE #771: add pip check migrator
Cross tracked issue from https://github.com/conda-forge/conda-forge.github.io/issues/962

### PR DESCRIPTION
**Pull request by beckermr**
I have started on the pip check migration.

To do:
- [x] handle outputs case where top-level name and output have the same name
- [x] deal with duplicate keys and selectors
- [x] deal with 'requires' versus 'requirements' in test sections :/
- [x] finish test suite
- [x] windows
- [x] piggyback on `Version` migrator in main code

closes #771 

cc @xhochy @isuruf @jakirkham @CJ-Wright @marcelotrevisani

### CODE REVIEW

**Comment by beckermr** [comment]
One way to fix the selectors would be to generate a diff by round-tripping the yaml first, adjust the recipe in memory, and then apply the diff after. This might work in a lot of cases, but won't be 100% robust.

File: conda_forge_tick/migrators/pip_check.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    if 'commands' in test:
+        test['commands'].append('{{ PYTHON }} -m pip check')
+    else:
+        test['commands'] = ['{{ PYTHON }} -m pip check']
+
+
+class PipCheckMigrator(MiniMigrator):
+    def filter(self, attrs: "AttrsTypedDict", not_bad_str_start: str = "") -> bool:
+        """run pip check if we see python in any host sections"""
+        host_reqs_list = (
```
Comment (CJ-Wright) [inline conda_forge_tick/migrators/pip_check.py]: You should use the `requirements` key in the attrs itself it already has all the things broken out into sets.

File: conda_forge_tick/migrators/pip_check.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            attrs.get("meta_yaml", {}).get("requirements", {}).get("host", [])
+        )
+        _, host_reqs = _has_python_in_host(host_reqs_list)
+
+        if "outputs" in attrs.get("meta_yaml", {}):
+            for output in attrs.get("meta_yaml", {})["outputs"]:
+                _host_reqs_list = output.get("requirements", {}).get("host", [])
+                _, _host_reqs = _has_python_in_host(_host_reqs_list)
+                host_reqs |= _host_reqs
+        return not bool(host_reqs & set(["python"]))
```
Comment (CJ-Wright) [inline conda_forge_tick/migrators/pip_check.py]: This should check for `pip` since not all packages use `pip`.

Comment (beckermr) [reply conda_forge_tick/migrators/pip_check.py]: pip check works for packages that don't use pip too IIUIC.

https://github.com/conda-forge/conda-forge.github.io/issues/962#issuecomment-579709883

File: conda_forge_tick/migrators/pip_check.py (hunk 1 above)
Comment (beckermr) [reply conda_forge_tick/migrators/pip_check.py]: Does this key handle outputs too? I don't actually know what is in the attrs object.

File: conda_forge_tick/migrators/pip_check.py (hunk 2 above)
Comment (CJ-Wright) [reply conda_forge_tick/migrators/pip_check.py]: yes, but I imagine it doesn't work if pip isn't installed.

File: conda_forge_tick/migrators/pip_check.py (hunk 1 above)
Comment (CJ-Wright) [reply conda_forge_tick/migrators/pip_check.py]: Yes, it also handles strong exports

Comment (CJ-Wright) [reply conda_forge_tick/migrators/pip_check.py]: https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/make_graph.py#L120

Comment (beckermr) [reply conda_forge_tick/migrators/pip_check.py]: Where can I find the code that gives the layout of this data?

I don't think I need strong exports for this migrator but that is awesome!

File: conda_forge_tick/migrators/pip_check.py (hunk 2 above)
Comment (beckermr) [reply conda_forge_tick/migrators/pip_check.py]: pip has to be in the test section, which is added by the code. It does not need to be in the host section.

File: conda_forge_tick/migrators/pip_check.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+def _has_python_in_host(host):
+    host_set = {r.split(" ")[0] for r in host}
+    return bool(host_set & set(["python"])), host_set
+
+
+def _adjust_test_dict(test):
+    if 'requires' in test:
+        test['requires'].append('pip')
+    else:
+        test['requires'] = ['pip']
```
Comment (beckermr) [inline conda_forge_tick/migrators/pip_check.py]: @CJ-Wright I am adding pip in the test requirements here.

File: conda_forge_tick/migrators/pip_check.py (hunk 1 above)
Comment (CJ-Wright) [reply conda_forge_tick/migrators/pip_check.py]: That block should give you the layout, it is a dict of sets.

File: conda_forge_tick/migrators/pip_check.py [hunk 4]
```diff
@@ -29,7 +29,9 @@ def _adjust_test_dict(test):
 class PipCheckMigrator(MiniMigrator):
     def filter(self, attrs: "AttrsTypedDict", not_bad_str_start: str = "") -> bool:
         """run pip check if we see python in any host sections"""
-        return "python" not in attrs['requirements']['host']
+        filter_top_level = "python" not in attrs['requirements'].get('host', set())
+        # FIXME - check outputs too
```
Comment (CJ-Wright) [inline conda_forge_tick/migrators/pip_check.py]: The requirements should have the outputs reqs in it as well.

**Comment by beckermr** [comment]
OK Team, some notes on this one since it is a beast.

1. It appears that the `PYTHON` jinja2 variable `{{ PYTHON }}` is not defined when the recipe is rendered for testing. IDK why exactly this happens or if this is a bug. In any case, in order to support both windows and unix, I am adding a `TEST_PYTHON` jinja2 variable at the top via `set` that takes on the correct value for each platform. This is pretty ugly but IDK what else to do that is robust.

2. The extensive munging of keys and various generators are there to support selectors. I carry them over to various inserted sections in order to attempt to generate valid recipes. 

   For example, if you have a block like this

   ```yaml
   test:  # [win]
    imports: # [win]
     - foo  # [win]
   ```
   then you need to add the trailing selector to any extra blocks you insert

   ```yaml
   test:  # [win]
    imports: # [win]
     - foo  # [win]
     requires:  # [win]
      - pip   # [win]
   ```
   The same logic/idea applies to appending to list blocks with selectors applied to the top-level key as well.
   
3. I was careful to add tests to recipes with more than one output where only one of the outputs might be a python output. I also tried to catch the case where the recipe has top-level tests that are run for a python output, but there is not `python` dep in the top-level `host` section by looking for and `imports` section in the tests.

One of the results of this PR is actually a general method to parse raw recipes in yaml via munging keys. This works, but I suspect @marcelotrevisani will have a better solution coming. :)

File: tests/test_yaml/version_pipcheck_outputs_correct.yaml [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+test:
+  imports:
+    - camb
+  files:
+    - params.ini
+    - test_pycamb.py
+    - run_test.sh  # [win]
+  commands:  # [not win]
+    - echo '1'  # [not win]
+    - '{{ TEST_PYTHON }} -m pip check'                                          # [not win]
```
Comment (isuruf) [inline tests/test_yaml/version_pipcheck_outputs_correct.yaml]: Can't you just do, `python -m pip check`?

Comment (beckermr) [reply tests/test_yaml/version_pipcheck_outputs_correct.yaml]: I thought it was python.exe on windows?

File: conda_forge_tick/auto_tick.py [hunk 6]
```diff
@@ -62,7 +63,10 @@
 
 
 MIGRATORS: MutableSequence[Migrator] = [
-    Version(pr_limit=30, piggy_back_migrations=[PipMigrator(), LicenseMigrator()]),
+    Version(
```
Comment (CJ-Wright) [inline conda_forge_tick/auto_tick.py]: Do you also want to put this on the other migrators as well?

File: conda_forge_tick/migrators/pip_check.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+                '# [%s]' % parent_group[3],
+                'commands',
+                column=80,
+            )
+
+
+class PipCheckMigrator(MiniMigrator):
+    def filter(self, attrs: "AttrsTypedDict", not_bad_str_start: str = "") -> bool:
+        """run pip check if we see python in any host sections"""
+        return "python" not in attrs['requirements'].get('host', set())
```
Comment (CJ-Wright) [inline conda_forge_tick/migrators/pip_check.py]:
```suggestion
        return "python" not in attrs['requirements'].get('host', set()) or attrs['requirements'].get('build', set()) or set()
```
This way we handle super old syntax where packages used build not host. (we should update that syntax but that might be out of scope for this PR)

File: tests/test_yaml/version_pipcheck_outputs_correct.yaml (hunk 5 above)
Comment (beckermr) [reply tests/test_yaml/version_pipcheck_outputs_correct.yaml]: Nevermind! This works on windows too! Thanks!

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=118793&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=7225e372-3a5e-5b11-c153-719d50bbeacd&l=967

**Comment by beckermr** [comment]
Alright, team! I think this one is ready to go!

cc @CJ-Wright @isuruf

**Comment by isuruf** [comment]
One issue would be that we'll not be checking that the package works without pip or setuptools. For eg: packages using `pkg_resources` needs setuptools as a run dependency, but this would not fail as pip is in test/requires.

**Comment by beckermr** [comment]
Good point! IDK what to do about this. :/

We cannot run `pip check` without `pip`. 

Further, `pip` doesn't always have complete metadata for packages anyways. Thus it is not clear this specific thing is the biggest issue.

**Comment by CJ-Wright** [comment]
You could add pip check as a script that installs pip and uninstalls it

**Comment by beckermr** [comment]
That doesn't solve the issue. Once you install `pip`, it will pull in `pkg_resources` and so you won't be able to check for this dep properly.

**Comment by jakirkham** [comment]
While I agree that would be nice to verify, I don't think that should be a blocker. Given the tooling we have today, this is a very good first step.

Perhaps we need to add a check like this to conda-build itself in the long-term.

**Comment by beckermr** [comment]
I suspect most actual conda envs have pip in them anyways, so this is better than nothing. 

Shall we merge @CJ-Wright @isuruf @jakirkham ?

**Comment by isuruf** [comment]
I would wait a  little bit to see how `pypy` turns out and see if we need a migration there.

**Comment by beckermr** [comment]
I don't follow that comment @isuruf. Why would we wait? 

We can piggyback this on that migration too if we want when that one gets submitted in the future (if that is your concern).

**Comment by isuruf** [comment]
Ah, so this is a piggyback migration?

**Comment by beckermr** [comment]
To be clear, this is a piggyback migration so it simply runs on top of other migrations. This is not a separate migration that generates its own PRs. 

So for now, mostly this change will appear in bot version update PRs and maybe some of the GSL and py38 stuff.

**Comment by jakirkham** [comment]
IOW a PyPy migration would also get this (once such a migration showed up). Is that right?

**Comment by beckermr** [comment]
indeed.  

However, we have been chatting offline and are thinking it might not make sense to use this at all. 

Here is our first pip check PR: https://github.com/conda-forge/sagemaker_containers-feedstock/pull/9

Lots of issues:

1. it says we need typing but that is only for python 2
2. the pypi metadata has a bunch of exact pins that the conda-forge recipe doesn't

IDK how much we actually want to gate PRs on this. The quality of the information seems rather poor.

(BTW, I put in a fix for the weird line wrapping.)

**Comment by jakirkham** [comment]
Well that raises another question that we need not answer here, but when do we want to drop Python 2?

**Comment by jakirkham** [comment]
As item 2, that may actually be a valid problem depending on whether the `setup.py` file also contained those pins.

**Comment by beckermr** [comment]
I think it has to right? That is the only place pypi would get the info IIUIC.

I think many maintainers are quietly ignoring pins like this from pypi. I actually think this might be an ok situation.

**Comment by beckermr** [comment]
@jakirkham what if we have the linter bot install packages as we lint PRs and run pip check there? It could then report any issues as linting hint.

**Comment by beckermr** [comment]
Granted this would be waaaay to slow for the current heroku setup.

**Comment by jakirkham** [comment]
Sounds like an interesting idea. Does `pip` use a solver for this? If not, it's probably really fast 😄

**Comment by beckermr** [comment]
pip does not have a true dependency solver so no. I think it just looks and compares for each requirement. Building the conda env with the package will be the slow part.

**Comment by jakirkham** [comment]
Right. Good point. Wasn't thinking about that.

**Pull request merged**